### ISSUE TITLE
Cannot compute gradient of functions that have default arguments

### ISSUE THREAD

**Issue by parth-07** [vgvassilev/clad#393]
Minimum reproducible example: 

```cpp
double fn(double i, double j = 0) {
  return i+j;
}

int main() {
  auto fn_grad = clad::gradient(fn);
}
```

**Output:**
```
error: missing default argument on parameter '_d_i'
error: missing default argument on parameter '_d_j'
2 errors generated. 
```

The root cause of the error is that we are adding default arguments in the derived function for the parameters that had default arguments in the original function. This does not work well for the derived functions because we append derivative parameters at the end, and C++ requires all parameters on the right-hand side of a parameter with default argument to have default argument as well. 
This can be easily fixed by removing default arguments in the derived function.

**Comment by Nirhar** [comment]
Hey I am new to this repository and would like to take up this issue! Can this issue be assigned to me?

After having checked out the codebase, I found that the parameter list for the reverse-mode derived functions are generated in the the Function, "OverloadedDeclWithContext  ReverseModeVisitor::Derive(const FunctionDecl* FD,  const DiffRequest& request)", especially in this section of the code:
```
    llvm::SmallVector<ParmVarDecl*, 4> params;
    llvm::SmallVector<ParmVarDecl*, 4> outputParams;
    params.reserve(m_Function->getNumParams() + args.size() + numExtraParam);
    for (auto* PVD : m_Function->parameters()) {
      Expr* PVDDefArg =
          PVD->hasDefaultArg() ? Clone(PVD->getDefaultArg()) : nullptr;
      auto VD =
          ParmVarDecl::Create(m_Context, gradientFD, noLoc, noLoc,
                              PVD->getIdentifier(), PVD->getType(),
                              PVD->getTypeSourceInfo(), PVD->getStorageClass(),
                              // Clone default arg if present.
                              PVDDefArg);
      if (VD->getIdentifier())
        m_Sema.PushOnScopeChains(VD, getCurrentScope(),
                                 /*AddToContext=*/false);
      params.push_back(VD);
```

Here we are just iterating through all the parameters and adding it to our parameter list, but also we check if the parameter already has a default argument or no. In my opinion, all we need to do is to change this portion so that the check for default arguments is disabled. Am I right with this? Shall I go ahead with fixing this issue or are there any other factors that I must consider?

**Comment by parth-07** [comment]
Yes, you are on the right track. But we also need to remove the default argument from the actual gradient and jacobian functions and not just from the overloaded derived function. And also, we seem to have the same problem in the hessian functions as well. 

This being said, this issue is already being fixed in the PR #397 . Can you try to resolve some other of the open issues (Maybe issue #388).

**Comment by Nirhar** [comment]
Yeah sure let me try issue #388
